pkg/scheduler/plugins/gang/gang.go

```diff
@@ -0,0 +1,437 @@
+package gang
+
+import (
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"encoding/json"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/scheduler-plugins/apis/scheduling/v1alpha1"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingconfig "github.com/koordinator-sh/koordinator/apis/scheduling/config"
+)
+
+var (
+	timeNowFn = time.Now
+)
+
+const (
+	GangFromCrd        string = "GangFromCrd"
+	GangFromAnnotation string = "GangFromAnnotation"
+)
+
+// Gang  basic gang info recorded in gangCache:
+type Gang struct {
+	Name       string
+	WaitTime   time.Duration
+	CreateTime time.Time
+	//strict-mode or non-strict-mode
+	Mode              string
+	MinRequiredNumber int
+	TotalChildrenNum  int
+	GangGroup         []string
+	Children          map[string]*v1.Pod
+	//pods that have already assumed(waiting in Permit stage)
+	WaitingForBindChildren map[string]*v1.Pod
+	//pods that have already bound
+	BoundChildren map[string]*v1.Pod
+	//if assumed  pods number has reached to MinRequiredNumber
+	ResourceSatisfied bool
+
+	//if the gang should be passed at PreFilter stage(Strict-Mode)
+	ScheduleCycleValid bool
+	//these fields used to count the cycle
+	ScheduleCycle            int
+	ChildrenScheduleRoundMap map[string]int
+
+	GangFrom    string
+	HasGangInit bool
+
+	lock sync.Mutex
+}
+
+func NewGang(gangName string) *Gang {
+	return &Gang{
+		Name:                     gangName,
+		CreateTime:               timeNowFn(),
+		WaitTime:                 0,
+		Mode:                     extension.GangModeStrict,
+		Children:                 make(map[string]*v1.Pod),
+		WaitingForBindChildren:   make(map[string]*v1.Pod),
+		BoundChildren:            make(map[string]*v1.Pod),
+		ScheduleCycleValid:       true,
+		ScheduleCycle:            1,
+		ChildrenScheduleRoundMap: make(map[string]int),
+		GangFrom:                 GangFromAnnotation,
+		HasGangInit:              false,
+	}
+}
+
+/******************Pod\PodGroup Event BEG******************/
```

Comment:
Please do not use this style of comments.

Reply:
ok